udf-compiler/src/main/scala/com/nvidia/spark/udf/CFG.scala

```diff
@@ -0,0 +1,328 @@
+/*
+ * Copyright (c) 2019-2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.udf
+
+import CatalystExpressionBuilder.simplify
+import javassist.bytecode.{CodeIterator, ConstPool, InstructionPrinter, Opcode}
+import scala.annotation.tailrec
+import scala.collection.immutable.{HashMap, SortedMap, SortedSet}
+
+import org.apache.spark.SparkException
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.catalyst.expressions._
+
+/**
+ * Control Flow Graph (CFG)
+ *
+ * This file defines the basic blocks (BB), and a class that can generate a CFG
+ * by leveraging [[LambdaReflection]].
+ */
+
+/**
+ * A Basic Block (BB) is a set of instructions defined by [[instructionTable]],
+ * where each entry in this table is a mapping of offset to [[Instruction]].
+ *
+ * The case class also provides some helpers, most importantly the [[propagateState]] helper
+ * which generates a map of [[BB]] to [[State]].
+ *
+ * @param instructionTable
+ */
+case class BB(instructionTable: SortedMap[Int, Instruction])  extends Logging {
+  def offset: Int = instructionTable.head._1
+
+  def last: (Int, Instruction) = instructionTable.last
+
+  def lastOffset: Int = last._1
+
+  def lastInstruction: Instruction = last._2
+
+  def propagateState(cfg: CFG, states: Map[BB, State]): Map[BB, State] = {
+    val state@State(_, _, cond, expr) = states(this)
+    logDebug(s"[BB.propagateState] propagating condition: ${cond} from state ${state} " +
+      s"onto states: ${states}")
+    lastInstruction.opcode match {
+      case Opcode.IF_ICMPEQ | Opcode.IF_ICMPNE | Opcode.IF_ICMPLT |
+           Opcode.IF_ICMPGE | Opcode.IF_ICMPGT | Opcode.IF_ICMPLE |
+           Opcode.IFLT | Opcode.IFLE | Opcode.IFGT | Opcode.IFGE |
+           Opcode.IFEQ | Opcode.IFNE | Opcode.IFNULL | Opcode.IFNONNULL => {
+        println(s"[BB.propagateState] lastInstruction: ${lastInstruction.instructionStr}")
```

Comment:
We should remove all `println`. They should likely become `logTrace`, and some be removed.

Reply:
Updated.